com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/pipeline/ShaderIncludeCompiler.java

```diff
@@ -0,0 +1,207 @@
+package com.dynamo.bob.pipeline;
+
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.regex.Matcher;
+import java.io.IOException;
+import java.io.File;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+
+import com.dynamo.bob.CompileExceptionError;
+import com.dynamo.bob.Project;
+import com.dynamo.bob.fs.IResource;
+import com.dynamo.bob.pipeline.ShaderUtil.Common;
+
+/* ShaderIncludeCompiler
+ * ========================
+ * This class contains functionality to support #include "path-to-file" directives from shader files.
+ * Includes must be in double quotes, <> and '' are currently not supported, and the included file
+ * must have the .glsl extension since the snippets must be picked up by bob during the build process.
+ * During tree creation, the tree will be checked for cycles since we will end up in an endless loop otherwise
+ * when building the tree.
+ *
+ * Relative and absolute paths are supported and must be project relative, examples:
+ *
+ * #include "/path/to/absolute-file.glsl"
+ * #include "file-in-same-dir.glsl"
+ * #include "path/to/sub-folder-file.glsl"
+ * #include "../file-in-parent-dir.glsl"
+ *
+ */
+public class ShaderIncludeCompiler {
+    // Compiler state
+    private Project     project;
+    private String      sourcePath;
+    private IncludeNode root;
+
+    private class IncludeNode {
+        public String                       path;
+        public String                       source;
+        public IncludeNode                  parent;
+        public HashMap<String, IncludeNode> children = new HashMap<String, IncludeNode>();
```

Comment:
Does the children have to be ordered?
Is it currently possible that the includes
```
#include "a.glsl"
#include "b.glsl"
```
WIll be added as:
```
float function_B() {...}
float function_A() {...}
```
?

Reply:
Ah right, I'm a dummy. Yes order needs to be preserved! I'll fix it 👍